File: benchmarks/tsb/bench_timedelta_arithmetic_fn.ts
/**
 * Benchmark: Timedelta arithmetic — add / subtract / abs / scale / comparisons.
 * Tests Timedelta construction via new Timedelta(ms) and its methods.
 * Outputs JSON: {"function": "timedelta_arithmetic_fn", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Timedelta } from "../../src/index.ts";

const WARMUP = 5;
const ITERATIONS = 100;

const SIZE = 1_000;
const td1 = new Timedelta(5_400_000); // 1.5 hours
const td2 = new Timedelta(1_800_000); // 30 minutes

const deltas = Array.from({ length: SIZE }, (_, i) =>
  new Timedelta((i - SIZE / 2) * 60_000),
);

for (let w = 0; w < WARMUP; w++) {
  for (const td of deltas.slice(0, 50)) {
    td.add(td1);
    td.subtract(td2);
    td.abs();
    td.scale(2);
    td.lt(td1);
    td.gt(td2);
  }
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const t0 = performance.now();
  for (const td of deltas) {
    td.add(td1);
    td.subtract(td2);
    td.abs();
    td.scale(2);
    td.lt(td1);
    td.gt(td2);
  }
  times.push(performance.now() - t0);
}

const total = times.reduce((a, b) => a + b, 0);
console.log(
  JSON.stringify({
    function: "timedelta_arithmetic_fn",
    mean_ms: round3(total / ITERATIONS),
    iterations: ITERATIONS,
    total_ms: round3(total),
  }),
);

function round3(v: number): number {
  return Math.round(v * 1000) / 1000;
}

File: benchmarks/pandas/bench_timedelta_arithmetic_fn.py
"""
Benchmark: pandas Timedelta add/sub/abs — basic Timedelta arithmetic.
Mirrors tsb bench_timedelta_arithmetic_fn.ts.
Outputs JSON: {"function": "timedelta_arithmetic_fn", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import pandas as pd

WARMUP = 5
ITERATIONS = 100

SIZE = 1_000
td1 = pd.Timedelta(days=1, hours=6)
td2 = pd.Timedelta(hours=2, minutes=30)
ms_value = 7_200_000  # 2 hours in ms

deltas = [pd.Timedelta(milliseconds=(i - SIZE // 2) * 60_000) for i in range(SIZE)]

for _ in range(WARMUP):
    pd.Timedelta(milliseconds=ms_value)
    for td in deltas[:50]:
        td + td1
        td - td2
        abs(td)

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    pd.Timedelta(milliseconds=ms_value)
    for td in deltas:
        td + td1
        td - td2
        abs(td)
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
print(json.dumps({
    "function": "timedelta_arithmetic_fn",
    "mean_ms": round(total_ms / ITERATIONS, 3),
    "iterations": ITERATIONS,
    "total_ms": round(total_ms, 3),
}))
